nixos/modules/services/networking/unbound.nix

```diff
@@ -6,41 +6,113 @@ let
+        };
+        description = "server: clause configuration";
+      };
+      remote-control = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
+          control-enable = false;
+          control-interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
+          server-key-file = "${cfg.stateDir}/unbound_server.key";
+          server-cert-file = "${cfg.stateDir}/unbound_server.pem";
+          control-key-file = "${cfg.stateDir}/unbound_control.key";
+          control-cert-file = "${cfg.stateDir}/unbound_control.pem";
+        };
+        description = "remote-control: clause configuration";
+      };
+      stub-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        description = "stub-zone: clause configuration";
+      };
+      forward-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        example = literalExample ''
+          [
+            {
+              name = ".";
+              forward-addr = "1.1.1.1@853#cloudflare-dns.com";
+            }
+            {
+              name = "example.org.";
+              forward-addr = [
+                "1.1.1.1@853#cloudflare-dns.com"
+                "1.0.0.1@853#cloudflare-dns.com"
+              ];
+            }
+          ];
+        '';
+        description = "forward-zone: clauses configuration";
+      };
+      python = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = { };
+        description = "python: clause configuration";
+      };
+    };
+  };
+
+  yesOrNo = v: if v then "yes" else "no";
+
+  toOption = n: v: "${toString n}: ${v}";
+
+  toConf = n: v:
+    if builtins.isFloat v then (toOption n (toString v))
+    else if isInt v       then (toOption n (toString v))
+    else if isBool v      then (toOption n (yesOrNo v))
+    else if isString v    then (toOption n v)
+    else if isList v      then (concatMapStringsSep "\n  " (toConf n) v)
+    else abort "unbound.toConf: unexpected type (n = ${n})";
+
+  clauseConf = clause: conf: if conf == { } then "" else
+    (concatStringsSep "\n  " (
+      ["${clause}:"] ++ (
+        mapAttrsToList toConf (
+          filterAttrs (_: v: v != null) conf
+        )
+      )));
+
+  confFile = pkgs.writeText "unbound.conf" (concatStringsSep "\n" (flatten [
+    (clauseConf "server" cfg.settings.server)
+    (clauseConf "remote-control" cfg.settings.remote-control)
+    (map (clauseConf "stub-zone") cfg.settings.stub-zone)
+    (map (clauseConf "forward-zone") cfg.settings.forward-zone)
+    (clauseConf "python" cfg.settings.python)
+  ]));
+
+  rootTrustAnchorFile = "${cfg.stateDir}/root.key";
+
+  unboundWrapped = pkgs.stdenv.mkDerivation {
```

Comment:
btw how about `pkgs.runCommand`?

Reply:
Is there a secret or something in the configuration file that requires this to live under `cfg.stateDir`?

Reply:
just briefly checked the config file on my server and at least there it doesn't. The `auto-trust-anchor-file` expects a filepath, so it should be fine.

Reply:
**If** there is the ability to specify all secrets as included files instead of directly embedding in the configuration file we should definitely avoid putting this file in the mutable state directory.

Reply:
So, is this okay?

Reply:
@rissson from what @Ma27 said it seems like there might be no hard requirement to store secrets in this file. Is that your understanding as well? If there is no requirement to store secrets in this file then we should not keep a copy of this file in the mutable state directory.

Reply:
I agree. Let's change this.

Reply:
Removed the config file from the `stateDir`